### DISCUSSION TITLE
SPGLIB failures to identify space group (test on 200 000 CSD structures)

### DISCUSSION THREAD

**Discussion by husakm** [spglib/spglib#537] [General]
Hi

I would like to return to the topic already discussed some time ago.

I had processed by SPGLIB about 200 000 crystal structures from Cambridge Structural Database. I can potentially process the complete database (1.5 million structures) .... I was focused only on structures with potential importance for pharmacy.
Generally SPGLIB had in a few thousands examples correctly identified incorrect space group determination by original authors or not-necessary super-cell usage ...
In a few hundred situations SPGLIB failed to give any result returning the 0 error space group number. I had used 0.1 tolerance factor witch corresponds to the typical errors in non-hydrogen atoms positions as given by X-ray diffraction based structure refinement (I did not use hydrogen atoms in the check) ...

Is there anybody interested in such failures analysis ?
I can:
A) Give the CSD ref-code list or even extract all the problematic CIF files form CSD.
B) If you do not have CIF SPGLIB interface I can automatically generate the inputs as .C source files.

Analysis on my side have probably limited sense - I do not 100% understand how SPGLIB internally works.

**Comment by LecrisUT** [comment]
I believe this is more of an academic issue to study, not a software/library issue. For testing purposes, what is needed are:
- unit tests to make sure all components works as expected
- functional tests to make sure there are no regression issues, and all basic structures are identifiable. For that I believe the [python tests](https://github.com/spglib/spglib/tree/develop/test/functional/python) are meant to cover them for now, but I don't know how thorough these are

**Reply by husakm** [reply]
I am not sure a test on real-live examples with great practical consequences can be named "academic issue".
Let me describe the practical aspects:
I must run spglib with relatively high tolerance (0.1-0.2).
This tolerance reflect the typical error of atoms positions as obtained by experimental X-ray diffraction on single crystal.
My test shows spglib gives no result (return 0 spg number) even for structures with clearly wrong original symmetr given (typicaly sub group of true group).
I do not understand why this happen. If I go down with the tolerance value (0.01), spglib gives for this examples correct results (but fail to detect higer symetry of more displaced other structures).
Maybe an adaptive algorithm re-runing spglib with lower tolerance when a failure occure can be a solution ?

**Reply by LecrisUT** [reply]
For me it makes sense:
- too low tolerance: fails to find symmetry due to numerical error moving atoms outside of lattice
- too high tolerance: multiple lattices could be a solution. Extreme example would be a cubic and hexagonal lattice.

Can the error reporting be improved to show why it fails, for sure, but I didn't navigate the code enough to pin-point where to patch this.

As for the real-live examples, those are not adequate because it is not controlled compared to the set of structures + controlled random noisy. It also needs to be vetted/peer-reviewed which for me puts that into academic issue.

**Comment by husakm** [comment]
Thank you for your opinion. I will implement iterative tolerance adjustment on my side. Hopefuly it will handle all the "no result" cases ...

**Comment by husakm** [comment]
I was able to eliminate the SPGLIB failure to give any results by following  algorithm (in C# based on my SPGLIB C# interface):
   
```
            SPGLIB.SpglibDataset dataset = new SPGLIB.SpglibDataset();

            double final_precision = starting_precision;
            int n_loops = 0; // we count iteration loops to be able to change the precision

            // iterate until SPGLib find a solution with higer precision than default
            do
            {
                final_precision = starting_precision / Math.Pow(2, n_loops);
                SPGLIB.SpglibError result = SPGLIB.spg_get_dataset(ref dataset, lattice, position, types, num_atom, final_precision);
                n_loops++; 
            }
            while ((dataset.spacegroup_number == 0) && (final_precision > 0.00001)); // solution found or precision to high

```

This code iteratively go down with precision until any solution is found ...
I start with starting precision = 0.1 A
This code solves the problem in most cases ..
For 18 structures there was no result even for 0.00001 less tolerance.
17 from them were one of following group:  R 3, R-3, R 3 c, R 3 m

Is it possible spglib have some problem with the R 3  like symmetry ? Or maybe there is an issue in our code with this groups (because we must perform R 3 -> P1 conversion before spglib launch to feed spglib by all atoms ...

**Reply by LecrisUT** [reply]
Could you check against an ideal structure and then add some controlled random fluctuations? I don't know about the specifics of the R3 symmetry so I can't offer any insight on what is going on there. But if this is a relatively known and well understood issue that can be reproduced with the random fluctuation, maybe we should investigate having a special case for these. Let's get something reproducible first though.

**Comment by husakm** [comment]
I was able to identify the issue wit the non R3 based structure. It is total nonsence input - error in database. So I have 16 R 3 derived problematic structures. I had already tested that even going with tolerance almost to zero does not help
I will:
- check each of them manualy
- prepare ideal structures + add bias as you suggest

**Comment by husakm** [comment]
The results of my tests give clear answer:
SPGLIB is not able to detect in any case correctly R 3 derived structures space group, even from totally ideal data.
It works OK when the cell is described as hexagonal, fail when the cell is described as rhombohedral (you probably know hexagonal and 
rhombohedral cells are transferable).
Instead of detecting R3 for perfect R3 structure in rhombohedral setting it detects P1 (tolerance close to 0 is required).
I had reproduced the issue on  5 different structures, perfect symmetry was generated by BIOVA Materials Studio code.
Possible reason of the issue:
a) General bug in SPGLIB occurring in this situation
b) Wrong SPGLIB data pre-processing on our side specific for this situation
For testing on your side I had transformed a rhombohedral setting of a structure with CSD code PUYHUM01 (ideal R 3) from CIF to a C code suitable for testing on your side (this is the most simple case generating this issue I detected) . Code follows ...

```
//section  generated by checkCIF-DFT
int num_atom =25;
double lattice[3][3] = {
{8.20100021362305,-2.10044598579407,-2.10044598579407},
{0,7.92745447158813,-2.72945523262024},
{0,0,7.4427547454834}
};

double position[25][3] = {
{0.1492999792099,0.998300015926361,0.901199996471405},
{0.266299962997437,0.182100057601929,0.935899972915649},
{0.199300050735474,0.857200026512146,0.809599995613098},
{0.107399940490723,0.678699970245361,0.787999987602234},
{0.147300004959106,0.538999974727631,0.692200005054474},
{0.279899954795837,0.571600019931793,0.617999970912933},
{0.374899983406067,0.746399998664856,0.643599987030029},
{0.335500001907349,0.887499988079071,0.737800002098083},
{0.901199996471405,0.1492999792099,0.998300015926361},
{0.935899972915649,0.266299962997437,0.182100057601929},
{0.809599995613098,0.199300050735474,0.857200026512146},
{0.787999987602234,0.107399940490723,0.678699970245361},
{0.692200005054474,0.147300004959106,0.538999974727631},
{0.617999970912933,0.279899954795837,0.571600019931793},
{0.643599987030029,0.374899983406067,0.746399998664856},
{0.737800002098083,0.335500001907349,0.887499988079071},
{0.998300015926361,0.901199996471405,0.1492999792099},
{0.182100057601929,0.935899972915649,0.266299962997437},
{0.857200026512146,0.809599995613098,0.199300050735474},
{0.678699970245361,0.787999987602234,0.107399940490723},
{0.538999974727631,0.692200005054474,0.147300004959106},
{0.571600019931793,0.617999970912933,0.279899954795837},
{0.746399998664856,0.643599987030029,0.374899983406067},
{0.887499988079071,0.737800002098083,0.335500001907349},
{0.958599984645844,0.958599984645844,0.958599984645844},
};

int types[25]= {
7,6,6,6,6,6,6,6,7,6,6,6,6,6,6,6,7,6,6,6,6,6,6,6,14,
};
//end of section  generated by checkCIF-DFT
```
